### PR TITLE
Add RAFNode: resonate-and-fire neuron

### LINKED ISSUE #746: Feature: resonate-and-fire neuron (RAFNode)
### Summary

SpikingJelly's neuron zoo is all integrate-and-fire family (IF / LIF / PLIF / QIF / EIF / Izhikevich / KLIF / CUBA-LIF / …). There is no **resonate-and-fire** neuron — a subthreshold *resonator* rather than an integrator. I'd like to contribute one, and want a sign-off on scope/naming first per `CONTRIBUTING.md`.

### Why

Resonate-and-fire (Izhikevich, *"Resonate-and-fire neurons"*, Neural Networks 14 (2001) 883–894) gives a genuinely different primitive: a 2-D linear subthreshold system with an intrinsic frequency, so the neuron responds preferentially to inputs near that frequency and shows damped subthreshold oscillations / post-inhibitory rebound. It's the natural building block for frequency-selective front-ends and has had recent deep-learning interest for audio and long time-series (e.g. the 2024 "resonate-and-fire" SNN work; balanced-RF neurons).

### Model sketch

State is a 2-D vector `(u, v)` (equivalently a complex `z = u + i·v`). Subthreshold dynamics per step:

```
z <- z * exp((b + i·ω) · dt) + x        # decay rate b < 0, angular frequency ω
spike when  v  (the imaginary / chosen component)  crosses  v_threshold
```

`BaseNode` integration:

- two registered memories (`u`, `v`) or one complex `z`;
- `neuronal_charge` = the rotate-and-decay update above;
- `neuronal_fire` = surrogate threshold on `v` (or `|z|`);
- `neuronal_reset` = hard reset of the firing component (subtract / to `v_reset`) — the other component is left, which is what gives the rebound behaviour;
- single-step and multi-step `step_mode`, `torch` backend to start.

### Open questions for you

1. Naming — `RAFNode`? `ResonateFireNode`? consistent with the `*Node` convention.
2. Should `ω` and `b` be fixed args, or learnable like `ParametricLIFNode`'s `tau` (a "parametric RAF")?
3. Fire on the imaginary component, on `|z|`, or make it configurable?
4. Any objection to a complex-tensor internal state vs. two real memories? (complex is cleaner but interacts with AMP / some backends.)

If the design is acceptable I'll open a PR with the neuron, a functional-forward equivalence test in `test/activation_based/`, and bilingual docstrings in the house style.

*(proposal drafted with AI assistance)*

### PR DESCRIPTION
**Pull request by tritsystem**
Closes #746.

## What this adds

`RAFNode` — a resonate-and-fire neuron (Izhikevich, *Resonate-and-fire neurons*, Neural Networks 14 (2001) 883-894) — per the design agreed in #746:

- `RAFNode` as the class name
- fixed `b` (decay rate), `omega` (angular frequency), `dt` — not learnable
- fires on the imaginary/voltage component (`self.v`)
- two real-valued states, `self.u` and `self.v` — not a complex tensor
- a single, clearly documented reset rule: reuses `BaseNode`'s soft/hard reset on `self.v` only; `self.u` is **never** reset, which is what produces post-inhibitory rebound
- `'torch'` backend only

## Sub-threshold dynamics

The real-valued expansion of `z <- z * exp((b + i*omega)*dt) + x` for `z = u + i*v`:

```
u[t] = alpha*(u[t-1]*cos(theta) - v[t-1]*sin(theta)) + x[t]
v[t] = alpha*(u[t-1]*sin(theta) + v[t-1]*cos(theta))
```

where `alpha = exp(b*dt)`, `theta = omega*dt`.

## How it fits the current architecture

Follows the `single_step_functional_forward` / explicit-state contract described in your comment on #746: `RAFNode(BaseNode)` implements `single_step_functional_forward` on explicit `(v, u)` state, overrides `materialize_states` to shape the second state `u` (same pattern `AdaptBaseNode` already uses for its adaptation current `w`), and reuses `BaseNode`'s generic multi-step path as-is — no override needed.

`functional.raf_step` is the explicit state-transition function, matching the `qif_step` / `izhikevich_step` house pattern (bilingual CN/EN docstring, same call shape).

## What I checked

- `test_raf_step_matches_module` — `functional.raf_step` output matches `RAFNode(...)` forward, parametrized over `v_reset` and `detach_reset`.
- `test_raf_reset_only_touches_v` — a spike leaves `self.u` exactly as the charge step produced it (proves the reset rule is scoped to `v`).
- `test_raf_single_step_multi_step_equivalence` — `step_mode="s"` run one step at a time vs `step_mode="m"` run on the whole sequence produce identical spikes, `v_seq`, and final `(u, v)`.
- Manually verified the model does real physics, not just passes shape checks: driving a `RAFNode(omega=1.0)` at its resonant frequency builds ~40x more subthreshold amplitude than driving it off-resonance (`omega=2.5`); gradients flow through backprop.
- `uvx ruff==0.15.22 check .` and `format --check` both clean on the changed files.
- Full `test/activation_based/` suite: 1223 passed. The 68 failures in this environment (triton kernels, GPU monitor) are pre-existing — verified identical via `git stash` against an unmodified checkout; this machine has no triton and no CUDA-visible GPU.

## How did AI help

AI (Claude, via Claude Code) implemented the neuron and tests from the design agreed in #746, after reading `BaseNode`/`AdaptBaseNode`/`IzhikevichNode`/`QIFNode` to match the current architecture rather than the older `neuronal_charge`/`neuronal_fire` pattern in my original issue sketch. I reviewed the math, the reset semantics, and the test coverage, and ran the resonance/rebound/gradient checks above myself.

## Summary by CodeRabbit

- **New Features**
  - Added resonate-and-fire (RAF) neuron support through `RAFNode`.
  - Added RAF state transitions for single-step and multi-step processing.
  - Supports configurable thresholds, reset behavior, surrogate functions, and oscillator parameters.
  - Supports the Torch backend and sequence processing.

- **Documentation**
  - Added RAF neuron and functional API documentation.
  - Added the feature to the unreleased changelog.

- **Tests**
  - Expanded coverage for RAF reset behavior, including voltage state handling.

### CODE REVIEW

**Comment by AllenYolk** [comment]
Maintainer follow-up: I pushed a small documentation-only commit completing the public type annotations and RST docstring, adding `RAFNode`/`raf_step` to the API docs, and updating the changelog. The existing RAF tests and full Sphinx build pass locally, and CI is green. No implementation changes. Thanks!
